libfoundation/src/foundation-handler.cpp

```diff
@@ -90,10 +93,141 @@ const MCHandlerCallbacks *MCHandlerGetCallbacks(MCHandlerRef self)
+    uindex_t t_arg_index;
+    t_arg_index = 0;
+    t_value_result = nil;
+    for(t_arg_index = 0; t_arg_index < t_arity; t_arg_index++)
+    {
+        MCHandlerTypeFieldMode t_mode;
+        t_mode = MCHandlerTypeInfoGetParameterMode(t_signature, t_arg_index);
+        
+        MCTypeInfoRef t_type;
+        t_type = MCHandlerTypeInfoGetParameterType(t_signature, t_arg_index);
+        
+        if (t_mode != kMCHandlerTypeFieldModeIn)
+            abort();
+        
+        MCResolvedTypeInfo t_resolved_type;
+        if (!MCTypeInfoResolve(t_type, t_resolved_type))
+        {
+            MCErrorThrowGeneric(nil);
+            goto error_exit;
+        }
+        
+        if (MCTypeInfoIsForeign(t_resolved_type . type))
+        {
+            const MCForeignTypeDescriptor *t_descriptor;
+            t_descriptor = MCForeignTypeInfoGetDescriptor(t_resolved_type . type);
+            if (t_descriptor -> defined != nil &&
+                !t_descriptor -> defined(args[t_arg_index]))
+                t_value_args[t_arg_index] = MCValueRetain(kMCNull);
+            else
+            {
+                if (t_descriptor -> bridgetype != kMCNullTypeInfo)
+                {
+                    if (!t_descriptor -> doimport(args[t_arg_index], false, t_value_args[t_arg_index]))
+                        goto error_exit;
+                }
+                else
+                {
+                    if (!MCForeignValueCreateAndRelease(t_resolved_type . named_type, args[t_arg_index], (MCForeignValueRef&)t_value_args[t_arg_index]))
+                        goto error_exit;
+                }
+            }
+        }
+        else
+        {
+            t_value_args[t_arg_index] = MCValueRetain((MCValueRef)args[t_arg_index]);
+        }
+    }
+    
+    if (!MCHandlerInvoke(t_handler, t_value_args, t_arity, t_value_result))
+        goto error_exit;
+    
+    MCTypeInfoRef t_return_type;
+    t_return_type = MCHandlerTypeInfoGetReturnType(t_signature);
+    
+    MCResolvedTypeInfo t_resolved_return_type;
+    t_return_type = MCHandlerTypeInfoGetReturnType(t_signature);
+    if (!MCTypeInfoResolve(t_return_type, t_resolved_return_type))
+    {
+        MCErrorThrowGeneric(nil);
+        goto error_exit;
+    }
+    
+    if (t_resolved_return_type . named_type != kMCNullTypeInfo)
+    {
+        if (MCTypeInfoIsForeign(t_resolved_return_type . type))
+        {
+            const MCForeignTypeDescriptor *t_descriptor;
+            t_descriptor = MCForeignTypeInfoGetDescriptor(t_resolved_return_type . type);
+            if (!t_descriptor -> doexport(t_value_result, true, ret))
+                goto error_exit;
+        }
+        else
+        {
+            *(MCValueRef *)ret = t_value_result;
+            t_value_result = nil;
+        }
+    }
+    
+    return;
+    
+error_exit:
+    if (t_value_result != nil)
+        MCValueRelease(t_value_result);
+    for(uindex_t i = 0; i < t_arg_index; i++)
+        MCValueRelease(t_value_args[i]);
+}
```

Comment:
Is it possible to use `MCAutoValueRef` and `MCAutoValueRefArray` to avoid explicit cleanup rules?